base_layer/core/src/mempool/utx_pool/utx_pool_storage.rs

```diff
@@ -0,0 +1,165 @@
+    mempool::utx_pool::{PrioritizedTransaction, UTxPoolConfig, UTxPoolError, UTxPriority},
+    transaction::Transaction,
+    types::Signature,
+};
+use std::{
+    collections::{BTreeMap, HashMap},
+    convert::TryFrom,
+};
+
+/// UTxPool makes use of UTxPoolStorage to provide thread save access to its Hashmap and BTreeMap.
+pub struct UTxPoolStorage {
+    config: UTxPoolConfig,
+    utx_hmap: HashMap<Signature, PrioritizedTransaction>,
+    utx_tree: BTreeMap<UTxPriority, Signature>,
+}
+
+impl UTxPoolStorage {
+    /// Create a new UTxPoolStorage with the specified configuration
+    pub fn new(config: UTxPoolConfig) -> Self {
+        Self {
+            config,
+            utx_hmap: HashMap::new(),
+            utx_tree: BTreeMap::new(),
+        }
+    }
+
+    fn lowest_priority(&self) -> &UTxPriority {
+        self.utx_tree.iter().next().unwrap().0
+    }
+
+    fn remove_lowest_priority_utx(&mut self) {
+        if let Some((priority, sig)) = self.utx_tree.iter().next().map(|(p, s)| (p.clone(), s.clone())) {
+            self.utx_hmap.remove(&sig);
+            self.utx_tree.remove(&priority);
+        }
+    }
+
+    /// Insert a new transaction into the UTxPoolStorage. Low priority transactions will be removed to make space for
+    /// higher priority transactions. The lowest priority transactions will be removed when the maximum capacity is
+    /// reached and the new transaction has a higher priority than the currently stored lowest priority transaction.
+    pub fn insert_utx(&mut self, utx: Transaction) -> Result<(), UTxPoolError> {
+        if utx.body.kernels.len() == 0 {
+            return Err(UTxPoolError::NoKernelsFound);
+        }
+        let utx_key = utx.body.kernels[0].excess_sig.clone();
+        if !self.utx_hmap.contains_key(&utx_key) {
+            let prioritized_utx = PrioritizedTransaction::try_from(utx)?;
+            if self.utx_hmap.len() >= self.config.storage_capacity {
+                if prioritized_utx.priority < *self.lowest_priority() {
+                    return Ok(());
+                }
+                self.remove_lowest_priority_utx();
+            }
+
+            self.utx_tree.insert(prioritized_utx.priority.clone(), utx_key.clone());
+            self.utx_hmap.insert(utx_key, prioritized_utx);
+        }
+        Ok(())
+    }
+
+    /// Insert a set of new transactions into the UTxPoolStorage
+    pub fn insert(&mut self, utxs: Vec<Transaction>) -> Result<(), UTxPoolError> {
+        for utx in utxs.into_iter() {
+            self.insert_utx(utx)?;
+        }
+        Ok(())
+    }
+
+    /// Check if a transaction is stored in the UTxPoolStorage
+    pub fn contains_key(&self, tx_excess_sig: &Signature) -> bool {
+        self.utx_hmap.contains_key(tx_excess_sig)
+    }
+
+    /// Returns a set of the highest priority unconfirmed transactions, that can be included in a block
+    pub fn highest_priority_utxs(&self, n: usize) -> Result<Vec<Transaction>, UTxPoolError> {
```

Comment:
Actually, switching to `max_weight` wouldn't be a stretch right now, given the nice way you've constructed this function.

BTW -- if we set the weights properly, if you had to maximise fee/byte, all TXs would have ± equal priority.

Reply:
Is it possible to return a `Vec` of borrowed `Transactions`?

Reply:
Changed to weight and sorting of transactions by fee per byte. It will now return a `Vec<Arc<Transaction>>`

Reply:
LGTM